homeassistant/components/mqtt/mixins.py

```diff
@@ -340,8 +350,10 @@ async def async_discover(discovery_payload):
             )
             raise
 
-    async_dispatcher_connect(
-        hass, MQTT_DISCOVERY_NEW.format(domain, "mqtt"), async_discover
+    hass.data.setdefault(DATA_MQTT_DISCOVERY_DISPATCHERS, []).append(
```

Comment:
The name is super confusing because you also use it for to track reload events, also just return the unsub and let it be handled by `entry.async_on_unload`